Svc/SeqDispatcher/SeqDispatcher.cpp

```diff
@@ -0,0 +1,199 @@
+  if (this->m_entryTable[sequencerIdx].state !=
+      SeqDispatcher_CmdSequencerState::AVAILABLE) {
+    this->log_WARNING_HI_InvalidSequencer(static_cast<U16>(sequencerIdx));
+    return false;
+  }
+
+  // ok, good to use this sequencer
+  if (block == Fw::Wait::NO_WAIT) {
+    this->m_entryTable[sequencerIdx].state =
+        SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_NO_BLOCK;
+  } else {
+    this->m_entryTable[sequencerIdx].state =
+        SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_BLOCK;
+  }
+
+  this->m_sequencersAvailable--;
+  this->tlmWrite_sequencersAvailable(this->m_sequencersAvailable);
+  this->m_entryTable[sequencerIdx].sequenceRunning = fileName;
+
+  this->m_dispatchedCount++;
+  this->tlmWrite_dispatchedCount(this->m_dispatchedCount);
+  this->seqRunOut_out(sequencerIdx,
+                      this->m_entryTable[sequencerIdx].sequenceRunning);
+
+  return true;
+}
+
+void SeqDispatcher::seqStartIn_handler(
+    NATIVE_INT_TYPE portNum, //!< The port number
+    const Fw::StringBase& fileName //!< The sequence file name
+) {
+  FW_ASSERT(portNum >= 0 && portNum < SeqDispatcherSequencerPorts, portNum);
+  if (this->m_entryTable[portNum].state ==
+          SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_BLOCK ||
+      this->m_entryTable[portNum].state ==
+          SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_NO_BLOCK) {
+    // we were aware of this sequencer running a sequence
+    if (this->m_entryTable[portNum].sequenceRunning != fileName) {
+      // uh oh. m_sequencesRunning is wrong
+      // let's just update it to be correct. nothing we can do about
+      // it except raise a warning and update our state
+      this->log_WARNING_HI_UnexpectedSequenceStarted(static_cast<U16>(portNum), fileName, this->m_entryTable[portNum].sequenceRunning);
+      this->m_entryTable[portNum].sequenceRunning = fileName;
+    }
+  } else {
+    // we were not aware that this sequencer was running. ground must have
+    // directly commanded that specific sequencer
+
+    // update the state
+    this->m_entryTable[portNum].state =
+        SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_NO_BLOCK;
+    this->m_entryTable[portNum].sequenceRunning = fileName;
+    this->m_sequencersAvailable--;
+    this->tlmWrite_sequencersAvailable(this->m_sequencersAvailable);
+  }
+}
+
+void SeqDispatcher::seqDoneIn_handler(
+    NATIVE_INT_TYPE portNum,         //!< The port number
+    FwOpcodeType opCode,             //!< Command Op Code
+    U32 cmdSeq,                      //!< Command Sequence
+    const Fw::CmdResponse& response  //!< The command response argument
+) {
+  FW_ASSERT(portNum >= 0 && portNum < SeqDispatcherSequencerPorts, portNum);
+  if (this->m_entryTable[portNum].state !=
+          SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_BLOCK &&
+      this->m_entryTable[portNum].state !=
+          SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_NO_BLOCK) {
```

Comment:
This catches the case where the state enum is in an invalid state.  If we do this simplification, we probably should also do:
```
FW_ASSERT(this->m_entryTable[portNum].state.isValid());
```

Reply:
Note, that goes before the if.

Reply:
I'm not sure I understand how this catches an invalid state.  The condition checks if the state is Available.  It seems awkward to check if the state is not in every other state that is not Available.

Reply:
@garthwatney , I think @LeStarch is saying that the original way I did the check catches the case in which the state is invalid (i.e. none of the given options). That said, I do see how it would be more direct and readable to just check if it's available. Happy to make this change and also add the isValid check.